plugin/documents.py

```diff
@@ -917,6 +923,27 @@ def _update_stored_selection_async(self) -> Tuple[Optional[sublime.Region], bool
         self._stored_selection = selection
         return changed_first_region, True
 
+    def _format_on_paste(self) -> None:
+        self.purge_changes_async()
+        clipboard_text = sublime.get_clipboard()
+        sel = self.view.sel()
+        split_clipboard_text = clipboard_text.split('\n')
+        multi_cursor_paste = len(split_clipboard_text) == len(sel) > 1
+        restore_regions = []
+        # add regions to selection, in order for lsp_format_document_range to format those regions
+        for index, region in enumerate(sel):
+            restore_regions.append(region)
+            look_text = clipboard_text
+            if multi_cursor_paste:
+                look_text = split_clipboard_text[index]
+            found_region = self.view.find(look_text, region.end(), sublime.REVERSE | sublime.LITERAL)
+            if found_region:
+                sel.add(found_region)
+        self.view.run_command('lsp_format_document_range')
+        # restore selection how it was
+        sel.clear()
+        sel.add_all(iter(restore_regions))
```

Comment:
Maybe something like this (continuing from my last comment):

```suggestion
    def _format_on_paste(self) -> None:
        self.purge_changes_async()
        clipboard_text = sublime.get_clipboard()
        sel = self.view.sel()
        original_selection = list(sel)
        split_clipboard_text = clipboard_text.split('\n')
        multi_cursor_paste = len(split_clipboard_text) == len(sel) > 1
        pasted_text = clipboard_text
        formatting_regions = []  # type: List[sublime.Region]
        # add regions to selection, in order for lsp_format_document_range to format those regions
        for index, region in enumerate(sel):
            if multi_cursor_paste:
                pasted_text = split_clipboard_text[index]
            pasted_region = self.view.find(pasted_text, region.end(), sublime.REVERSE | sublime.LITERAL)
            if pasted_region:
                # Expand to preceding whitespace on the same line
                a = self.view.find_by_class(pasted_region.a, False,
                        sublime.CLASS_WORD_END | sublime.CLASS_PUNCTUATION_END | sublime.CLASS_LINE_START)
                formatting_region = sublime.Region(a, pasted_region.b)
                formatting_regions.append(formatting_region)
        sel.add_all(iter(formatting_regions))
        self.view.run_command('lsp_format_document_range')
        sel.clear()
        sel.add_all(iter(original_selection))
```

I couldn't figure out how to expand to the trailing whitespace on the same line, because `sublime.CLASS_LINE_END` doesn't seem to work when I tried it.